src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/CreateUserService.kt

```diff
@@ -0,0 +1,134 @@
+package uk.gov.justice.digital.hmpps.externalusersapi.service
+
+import com.microsoft.applicationinsights.TelemetryClient
+import kotlinx.coroutines.flow.count
+import kotlinx.coroutines.flow.toList
+import org.apache.commons.lang3.StringUtils
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+import uk.gov.justice.digital.hmpps.externalusersapi.config.AuthenticationFacade
+import uk.gov.justice.digital.hmpps.externalusersapi.repository.RoleRepository
+import uk.gov.justice.digital.hmpps.externalusersapi.repository.UserGroupCoroutineRepository
+import uk.gov.justice.digital.hmpps.externalusersapi.repository.UserRepository
+import uk.gov.justice.digital.hmpps.externalusersapi.repository.UserRoleCoroutineRepository
+import uk.gov.justice.digital.hmpps.externalusersapi.repository.entity.Group
+import uk.gov.justice.digital.hmpps.externalusersapi.repository.entity.User
+import uk.gov.justice.digital.hmpps.externalusersapi.repository.entity.UserGroup
+import uk.gov.justice.digital.hmpps.externalusersapi.repository.entity.UserRole
+import uk.gov.justice.digital.hmpps.externalusersapi.resource.CreateUser
+import uk.gov.justice.digital.hmpps.externalusersapi.security.AuthSource
+import java.util.UUID
+
+@Service
+@Transactional(readOnly = true)
+class CreateUserService(
+  private val userRepository: UserRepository,
+  private val authenticationFacade: AuthenticationFacade,
+  private val telemetryClient: TelemetryClient,
+  private val userGroupService: UserGroupService,
+  private val roleRepository: RoleRepository,
+  private val userGroupCoroutineRepository: UserGroupCoroutineRepository,
+  private val userRoleCoroutineRepository: UserRoleCoroutineRepository,
+) {
+  @Transactional
+  @Throws(CreateUserException::class, UserExistsException::class)
+  suspend fun createUserByEmail(createUser: CreateUser): UUID? {
+    log.info("Creating External User : {} ", createUser.email)
+    var user = userRepository.findByUsernameAndSource(StringUtils.upperCase(createUser.email))
+    if (user != null) {
+      throw UserExistsException("userId", "User with username ${createUser.email} already exist")
+    }
+
+    if (userRepository.findByEmailAndSourceOrderByUsername(createUser.email).count() >= 1) {
+      throw UserExistsException("email", "Email with username ${createUser.email} already exist")
+    }
+
+   /* This check not required as endpoint is  protected with 'MAINTAIN_OAUTH_USERS' role
+    if (!canMaintainUsers(authenticationFacade.getAuthentication().authorities)) {
+      throw CreateUserException("groupCode", "missing")
+    } */
+    val group = createUser.groupCodes?.let { getInitialGroups(it) }
```

Comment:
prob rename to groups to be more explicit

Reply:
updated to groups